rust/tw_evm/src/address.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use std::fmt::{Display, Formatter};
+use std::ops::{RangeFrom, RangeInclusive};
+use std::str::FromStr;
+use tw_coin_entry::coin_entry::CoinAddress;
+use tw_coin_entry::error::{AddressError, AddressResult};
+use tw_encoding::hex;
+use tw_hash::{sha3::keccak256, H160};
+use tw_keypair::ecdsa::secp256k1;
+use tw_memory::Data;
+
+pub trait EvmAddress: FromStr<Err = AddressError> + Into<Address> {
+    /// Tries to parse an address from the string representation.
+    /// Returns `Ok(None)` if the given `s` string is empty.
+    #[inline]
+    fn from_str_optional(s: &str) -> AddressResult<Option<Self>> {
+        if s.is_empty() {
+            return Ok(None);
+        }
+
+        Self::from_str(s).map(Some)
+    }
+}
+
+/// Represents an Ethereum address.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Address {
+    bytes: H160,
+}
+
+/// cbindgen:ignore
+impl Address {
+    pub const LEN: usize = 20;
+
+    /// Initializes an address with a `secp256k1` public key.
+    pub fn with_secp256k1_pubkey(pubkey: &secp256k1::PublicKey) -> Address {
+        const ADDRESS_HASH_RANGE: RangeFrom<usize> = 12..;
+
+        let pubkey_bytes = pubkey.uncompressed_without_prefix();
+        let hash = keccak256(pubkey_bytes.as_slice());
+        assert_eq!(hash.len(), 32);
```

Comment:
32 seems to be required hash size, can we wrap it into a constant? Actually I'm quite sure that H160:: have something for constant size no?

Reply:
I dream to refactor `keccak256` and other hash functions to return a strict `H256`, `H512` etc. So this check will be no longer required. But I agree with you that better to use `H256::LEN` here